tieredstore/src/main/java/org/apache/rocketmq/tieredstore/provider/TieredFileSegmentInputStream.java

```diff
@@ -0,0 +1,323 @@
+         */
+        private int curReadBufferIndex = 0;
+        /**
+         * readPosInCurBuffer is the position in the buffer which is being read
+         */
+        private int readPosInCurBuffer = 0;
+
+        private ByteBuffer curBuffer;
+
+        private final List<ByteBuffer> uploadBufferList;
+
+        private int markCurReadBufferIndex = -1;
+        private int markReadPosInCurBuffer = -1;
+
+
+        public ConsumeQueueInputStream(TieredFileSegment.FileSegmentType fileType, List<ByteBuffer> uploadBufferList, int contentLength) {
+            super(fileType, contentLength);
+            this.uploadBufferList = uploadBufferList;
+            if (uploadBufferList.size() > 0) {
+                this.curBuffer = uploadBufferList.get(0);
+            }
+        }
+
+        @Override
+        public int realRead() {
+            if (curReadBufferIndex >= uploadBufferList.size()) {
+                return -1;
+            }
+            if (readPosInCurBuffer >= curBuffer.remaining()) {
+                curReadBufferIndex++;
+                if (curReadBufferIndex >= uploadBufferList.size()) {
+                    return -1;
+                }
+                curBuffer = uploadBufferList.get(curReadBufferIndex);
+                readPosInCurBuffer = 0;
+            }
+            return curBuffer.get(readPosInCurBuffer++) & 0xff;
+        }
+
+        @Override
+        public void realMark() {
+            this.markCurReadBufferIndex = curReadBufferIndex;
+            this.markReadPosInCurBuffer = readPosInCurBuffer;
+        }
+
+        @Override
+        public void realReset() {
+            this.curReadBufferIndex = markCurReadBufferIndex;
+            this.readPosInCurBuffer = markReadPosInCurBuffer;
+            if (curReadBufferIndex < uploadBufferList.size()) {
+                this.curBuffer = uploadBufferList.get(curReadBufferIndex);
+            }
+        }
+
+        @Override
+        public List<ByteBuffer> getUploadBufferList() {
+            return this.uploadBufferList;
+        }
+
+        @Override
+        public ByteBuffer getCodaBuffer() {
+            return null;
+        }
+    }
+
+    private static class IndexInputStream extends TieredFileSegmentInputStream {
+
+        private final ByteBuffer curBuffer;
+
+        private int readPosInCurBuffer = 0;
+
+        private int markReadPosInCurBuffer = -1;
+
+        public IndexInputStream(TieredFileSegment.FileSegmentType fileType, ByteBuffer curBuffer, int contentLength) {
+            super(fileType, contentLength);
+            if (curBuffer == null) {
+                throw new IllegalArgumentException("curBuffer is null");
+            }
+            this.curBuffer = curBuffer;
+        }
+
+        @Override
+        public int realRead() {
+            if (readPosInCurBuffer >= curBuffer.remaining()) {
+                return -1;
+            }
+            return curBuffer.get(readPosInCurBuffer++) & 0xff;
```

Comment:
Are you assuming IndexFile has only one buffer? It is better to reuse the logic of consume queue input stream